### ISSUE TITLE
cannot download in webview

### ISSUE THREAD

**Issue by tcurdt** [OpenRoberta/robertalab#627]
We have OpenRoberta running within a webview on iOS (`WKWebView`) and we would like to intercept and receive the downloads inside the app. While most other editors provide a normal or base64 encoded URL for the file, OpenRoberta provides a URL in the form of `blob:https://...` which is not something we can download from inside the app.

**Comment by ensonic** [comment]
If the code is open-source, getting a link would be useful...
What download are you trying to intercept - exporting the project? Also when saying 'most other editors' - do you have an example?

**Comment by tcurdt** [comment]
@ensonic sorry, no link just yet. On every navigation action a webview will get a callback with the request. If it's a hex file the navigation can be canceled and the download can/should happen outside of the webview.

```
    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {

        if let download = editor.download(navigationAction.request) {

            // it's a hex download, cancel the navigation action
            decisionHandler(.cancel)

            // now download the url that has the hex file
            print(download.url.absoluteString)
```
The [microbit editor](https://makecode.microbit.org/) sends the data as base64 encoded URL.
The [calliope mini editor](https://miniedit.calliope.cc/) sends the data as a direct download.

With OpenRoberta I am getting an URL to a blob that (I assume) is internal to the webview and I don't see a way to get access to the data.

**Comment by bjost2s** [comment]
I easily could remove the blob, but how can I test it?
As we are close to a release I can only commit this in a separate branch. Could you set up a server on your own and test it?

**Comment by tcurdt** [comment]
@bjost2s I haven't looked into the details of the deployment procedures but I am pretty certain I should be able to test it if it is on a branch.

**Comment by bjost2s** [comment]
I have checked the actual version from our lab again. Blobs only appear when downloading xml files. For Calliope you get a modal with the following html:
`<div id="downloadLink"><div id="programLink" style="text-align: center;"><br><a style="font-size:36px; padding: 20px" download="NEPOprog.hex" href="data:Application/octet-stream;content-disposition:attachment;charset=utf-8,%3A020 ... ">NEPOprog</a>`
where ... is the full hex content.

What exactly should I improve

**Comment by tcurdt** [comment]
Please see the screencast:

![screencast](https://user-images.githubusercontent.com/13697/31085748-4d7dcff0-a798-11e7-88d3-44fa1dd859f8.gif)

I can see that the "NEPOprog" link is of the structure you are referring to and it should/would work - but of course one would want to check the box and just use the "run" button.

BTW: The two last points of the instructions don't really make much sense to me yet. "left column"? "save button"?

**Comment by bjost2s** [comment]
Our instructions are for desktop browser. If you right click on a Link you will have the possibility To choose in the left column the Drive an the save the program there . The user doesn' t have To leave the Lab. 
What happens in your screencast, if you simply click/touch on the link. The download process should start immediatly, without asking for drives etc. I think that would be the point where the wepapp Could grap the hex.

**Comment by tcurdt** [comment]
@bjost2s I didn't understand that the instruction were meant for the file save dialog :)

If I just click on the link it works fine. But as I said before: I want to check the box and just use the "run" button - and that's where things break. Because then I am just getting an url to a blob.

**Comment by bjost2s** [comment]
I pushed a possible solution to branch calliopeDownload. Please check it.

**Comment by bjost2s** [comment]
Sorry, forgot to remove the alert ;-), but most interesting thing is if your app could grap the hex now

**Comment by bjost2s** [comment]
Any comments? Is it working in general?
I would then start some test to release it as soon as possible.

**Comment by tcurdt** [comment]
@bjost2s from the diff it looks OK (at least on the first glance - given that I don't know the code base)

Unfortunately I will have to make time to work out the installation first to test this. Which means setting up a vm and the installing it. Or is there by any chance a docker image that should work?

So far I found https://hub.docker.com/r/exmatrikulator/robertalab/

**Comment by bjost2s** [comment]
what we have seen and tested, it will not work. I'll send you a PM

**Comment by tcurdt** [comment]
Docker: I am very certain that it can be done - but if you don't have it working it's a moo point for now.

Branch: Thanks for access to the server. I can confirm it worked for us!

Looking forward to see this merged into the next release. Thanks!

**Comment by bjost2s** [comment]
seems to work on ubuntu chrome and FF, windows 10 chrome and FF, safari?

**Comment by bjost2s** [comment]
It seems that you managed to grap the downloaded hex from the "blob", if so please close this, I will not merge it to develop.

**Comment by tcurdt** [comment]
@bjost2s wait - what? I confirmed the branch is working and we are waiting for a release with it.

**Comment by bjost2s** [comment]
sorry, someone told me that your new app is already working with the open roberta lab. 
I will merge it to develop

**Comment by tcurdt** [comment]
@bjost2s Ah, sorry for the confusion. Yes, we released the app but Open Roberta is hardly usable without the change. We have another editor that works - but it would be really great to get this in so Open Roberta can also be fully supported. We just couldn't wait any longer for the release. Thanks!

**Comment by bjost2s** [comment]
We have trouble with Edge and IE, so I add a fix for this browsers.

**Comment by bjost2s** [comment]
please check in published 2.5.0 version online. We had to do some changes, no idea if the fix is still working for IOS webview. Please reopen, when it is still an issue. (release is already published, online tomorrow morning)

**Comment by bjost2s** [comment]
not working as reported by callliope
